### PR TITLE
[SPARK-5763][Core]add Sort-Merge Join to resolve skewed data

### PR DESCRIPTION
**Pull request by lianhuiwang**
add sort-merge join to resolve skewed data.
i provide three interface to achieve join operator using SortMergeJoinRDD, there are:
sortMergeJoin, leftOuterSortMergeJoin and rightOuterSortMergeJoin.
their usages ara same as join, leftOuterJoin and rightOuterJoin.
@rxin @sryza

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/SortMergeJoinRDD.scala [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                                          L: ClassTag,
+                                          R: ClassTag,
+                                          PAIR <: Product2[_, _]]
+    extends Serializable {
+
+  private[spark] var joinType: Int = 0
+
+  def flatten(iterators: Iterator[(K, (Iterator[L], Iterator[R]))]): Iterator[(K, PAIR)]
+
+  def mergeIterators(leftIter: Iterator[L], rightIter: Iterator[R]): Iterator[(L, R)] = {
```
Comment (redbaron) [inline core/src/main/scala/org/apache/spark/rdd/SortMergeJoinRDD.scala]: Not saying that it is wrong, just curious, is it any different from ziping iterators?

**Comment by andrewor14** [comment]
@rxin @sryza @lianhuiwang any updates on this patch? Is this something we want?

**Comment by rxin** [comment]
@lianhuiwang can we close this? Sort-merge-join is now part of Spark SQL / DataFrame.

**Comment by lianhuiwang** [comment]
yes, i will close this. thanks.

**Pull request closed without merging**